tools/clang/lib/Sema/SemaTemplate.cpp

```diff
@@ -2880,7 +2880,7 @@ ExprResult Sema::BuildTemplateIdExpr(const CXXScopeSpec &SS,
   //       vs template<class T, class U> void f(U);
 
   // These should be filtered out by our callers.
-  assert(!R.empty() && "empty lookup results when building templateid");
+  assert((!R.empty() || RequiresADL) && "empty lookup results when building templateid");
```

Comment:
I don't think you should need this change.

Reply:
Omitting this change results in an internal assert error when compiling the code provided in the original issue. Tracing the stack shows that all asserts encountered prior to this one reflect the structure of ```assert((!R.empty() || RequiresADL) && ... );``` and the assert shown here was failing because it receives ```R.empty() == true && RequiresADL == true```, which was accepted by the previous asserts.

Reply:
I don't locally hit this assert with the code for the original issue or in any of our test cases.

Reply:
Just as a note: the reason I'm concerned about this change is because this isn't an assert written by the HLSL compiler team, it is an assert inherited from Clang (I think authored by John McCall). If we're hitting this assert as written, we're probably doing something else wrong and in for a world of pain.

Reply:
Ah I see! Well it was necessary when I initially wrote the bug fix, but it is possible that something else has been changed since then that resolves the issue otherwise. Especially if you no longer encounter the problem with the original code from the issue.